File: src/main/java/com/manabreakstudios/discourse/ui/core/utils/SwingUtils.java
package com.manabreakstudios.discourse.ui.core.utils;

import com.manabreakstudios.discourse.ui.core.editor.NodeUI;

import javax.swing.*;
import java.awt.*;
import java.util.Arrays;
import java.util.List;

public class SwingUtils {

    private static final List<Class> INPUT_TRANSPARENT_COMPONENTS = Arrays.asList(JLabel.class, JPanel.class, JLayeredPane.class);

    public static Point screenToLocal(Point screen, Component component) {
        Point result = new Point(screen);
        SwingUtilities.convertPointFromScreen(result, component);
        return result;
    }

    public static NodeUI getNodeUIAncestor(Component component) {
        Component ancestor = component;
        while (ancestor != null) {
            if (ancestor instanceof NodeUI) {
                return (NodeUI) ancestor;
            }
            ancestor = ancestor.getParent();
        }
        return null;
    }

    public static Component getComponentAt(Component parent, Point point) {
        return SwingUtilities.getDeepestComponentAt(parent, point.x, point.y);
    }

    public static boolean isInputTransparent(Component component) {
        for (Class aClass : INPUT_TRANSPARENT_COMPONENTS) {
            if (aClass.isAssignableFrom(component.getClass())) {
                return true;
            }
        }
        return false;
    }

    public static Rectangle expand(Rectangle rect, int size) {
        rect.setRect(rect.x - size, rect.y - size, rect.width + 2 * size, rect.height + 2 * size);
        return rect;
    }

    public static void drawCircle(Graphics2D g2d, int x, int y, int size) {
        g2d.fillRoundRect(x, y, size, size, size, size);
    }
}

File: node-editor/src/com/manabreak/node_editor/ui/SwingUtils.kt
package com.manabreak.node_editor.ui

import javax.swing.*
import java.awt.*
import java.util.Arrays

object SwingUtils {

    private val INPUT_TRANSPARENT_COMPONENTS = Arrays.asList(JLabel::class.java, JPanel::class.java, JLayeredPane::class.java)

    fun screenToLocal(screen: Point, component: Component): Point {
        val result = Point(screen)
        SwingUtilities.convertPointFromScreen(result, component)
        return result
    }

    fun getNodeUIAncestor(component: Component): NodeUI<*>? {
        var ancestor: Component? = component
        while (ancestor != null) {
            if (ancestor is NodeUI<*>) {
                return ancestor
            }
            ancestor = ancestor.parent
        }
        return null
    }

    fun getComponentAt(parent: Component, point: Point): Component {
        return SwingUtilities.getDeepestComponentAt(parent, point.x, point.y)
    }

    fun isInputTransparent(component: Component): Boolean {
        for (aClass in INPUT_TRANSPARENT_COMPONENTS) {
            if (aClass.isAssignableFrom(component.javaClass)) {
                return true
            }
        }
        return false
    }

    fun expand(rect: Rectangle, size: Int): Rectangle {
        rect.setRect((rect.x - size).toDouble(), (rect.y - size).toDouble(), (rect.width + 2 * size).toDouble(), (rect.height + 2 * size).toDouble())
        return rect
    }

    fun drawCircle(g2d: Graphics2D, x: Int, y: Int, size: Int) {
        g2d.fillRoundRect(x, y, size, size, size, size)
    }
}
